src/pool.rs

```diff
@@ -0,0 +1,253 @@
+    shard: &'a Shard<T, C>,
+    key: usize,
+}
+
+impl<T, C> Pool<T, C>
+where
+    T: Clear + Default,
+    C: cfg::Config,
+{
+    /// The number of bits in each index which are used by the pool.
+    ///
+    /// If other data is packed into the `usize` indices returned by
+    /// [`Pool::create`], user code is free to use any bits higher than the
+    /// `USED_BITS`-th bit freely.
+    ///
+    /// This is determined by the [`Config`] type that configures the pool's
+    /// parameters. By default, all bits are used; this can be changed by
+    /// overriding the [`Config::RESERVED_BITS`][res] constant.
+    ///
+    /// [`Config`]: trait.Config.html
+    /// [res]: trait.Config.html#associatedconstant.RESERVED_BITS
+    /// [`Slab::insert`]: struct.Slab.html#method.insert
+    pub const USED_BITS: usize = C::USED_BITS;
+
+    /// Creates a new object in the pool, returning a key that can be used to access it.
+    ///
+    /// If this function returns `None`, then the shard for the current thread is full and no items
+    /// can be added until some are removed, or the maximum number of shards has been reached.
+    ///
+    /// # Examples
+    /// ```rust
+    /// # use sharded_slab::Pool;
+    /// let pool: Pool<String> = Pool::new();
+    /// let key = pool.create_with(|item| item.push_str("Hello")).unwrap();
+    /// assert_eq!(pool.get(key).unwrap(), String::from("Hello"));
+    /// ```
+    pub fn create_with(&self, initializer: impl FnOnce(&mut T)) -> Option<usize> {
+        let tid = Tid::<C>::current();
+        let mut init = Some(initializer);
+        test_println!("pool: create {:?}", tid);
+        self.shards[tid.as_usize()]
+            .init_with(|slot| {
+                let init = init.take().expect("initializer will only be called once");
+                slot.initialize_state(init)
+            })
+            .map(|idx| tid.pack(idx))
+    }
+
+    /// Return a reference to the value associated with the given key.
+    ///
+    /// If the pool does not contain a value for the given key, `None` is returned instead.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use sharded_slab::Pool;
+    /// let pool: Pool<String> = sharded_slab::Pool::new();
+    /// let mut value = Some(String::from("hello world"));
+    /// let key = pool.create(move |item| *item = value.take().expect("crated twice")).unwrap();
+    ///
+    /// assert_eq!(pool.get(key).unwrap(), String::from("hello world"));
+    /// assert!(pool.get(12345).is_none());
+    /// ```
+    pub fn get(&self, key: usize) -> Option<PoolGuard<'_, T, C>> {
+        let tid = C::unpack_tid(key);
+
+        test_println!("pool: get{:?}; current={:?}", tid, Tid::<C>::current());
+        let inner = self.shards.get(tid.as_usize())?.get(key, |x| x)?;
+
+        Some(PoolGuard {
+            inner,
+            // Safe access as previous line checks for validity
+            shard: &self.shards[tid.as_usize()],
+            key,
+        })
```

Comment:
nit/style: i think we could make the same change here as we did for `Slab`, to avoid the second indexing:
```suggestion
        let shard = self.shards.get(tid.as_usize())?;
        let inner = shard.get(key, |x| x)?;

        Some(PoolGuard {
            inner,
            shard,
            key,
        })
```